src/launchpad/service.py

```diff
@@ -78,58 +127,396 @@ def handle_kafka_message(self, payload: PreprodArtifactEvents) -> None:
+                self._statsd.increment(
+                    "launchpad.artifact.processing.error",
+                    tags=[
+                        f"error_code:{error_code}",
+                        f"error_type:{error_message.name}",
+                        f"project_id:{project_id}",
+                        f"organization_id:{organization_id}",
+                    ],
+                )
+
+            result = sentry_client.update_artifact(
+                org=organization_id,
+                project=project_id,
+                artifact_id=artifact_id,
+                data={"error_code": error_code, "error_message": final_error_message},
+            )
+
+            if "error" in result:
+                logger.error(f"Failed to update artifact with error: {result['error']}")
+            else:
+                logger.info(f"Successfully updated artifact {artifact_id} with error information")
+
+        except Exception as e:
+            logger.error(f"Failed to update artifact {artifact_id} with error information: {e}", exc_info=True)
+
+    def _handle_download_error(
+        self, sentry_client: SentryClient, artifact_id: str, project_id: str, organization_id: str, error_msg: str
+    ) -> None:
+        """Handle download error by logging and updating artifact."""
+        logger.error(error_msg)
+        # Extract just the error details for the database, the enum prefix will be added automatically
+        detailed_error = error_msg.split(": ", 1)[1] if ": " in error_msg else error_msg
+        self._update_artifact_error(
+            sentry_client,
+            artifact_id,
+            project_id,
+            organization_id,
+            ERROR_CODE_ARTIFACT_PROCESSING_ERROR,
+            ProcessingErrorMessage.DOWNLOAD_FAILED,
+            detailed_error,
+        )
+        raise RuntimeError(error_msg)
+
+    def _download_artifact(
+        self,
+        sentry_client: SentryClient,
+        artifact_id: str,
+        project_id: str,
+        organization_id: str,
+    ) -> tuple[bytes, int]:
+        """Download artifact from Sentry and validate response."""
+        logger.info(f"Downloading artifact {artifact_id}...")
 
         try:
-            # Wait for shutdown signal
-            await self._shutdown_event.wait()
+            download_result = sentry_client.download_artifact(
+                org=organization_id, project=project_id, artifact_id=artifact_id
+            )
+
+            if "error" in download_result:
+                error_category, error_description = categorize_http_error(download_result)
+                self._handle_download_error(
+                    sentry_client,
+                    artifact_id,
+                    project_id,
+                    organization_id,
+                    f"Failed to download artifact ({error_category}): {error_description}",
+                )
+
+            if not download_result.get("success"):
+                self._handle_download_error(
```

Comment:
The control flow here seems odd, `_handle_download_error` ends up rasing iiuc. Can `download_artifact` just raise an error directly?